iothub/device/src/Transport/Mqtt/OrderedTwoPhaseWorkQueue.cs

```diff
@@ -46,11 +46,13 @@ public Task CompleteWorkAsync(IChannelHandlerContext context, TWorkId workId)
                 throw new IotHubException("Nothing to complete.", isTransient: false);
             }
 
-            IncompleteWorkItem incompleteWorkItem = _incompleteQueue.Peek();
-            if (incompleteWorkItem.Id.Equals(workId))
+            if (_incompleteQueue.TryPeek(out IncompleteWorkItem incompleteWorkItem)
```

Comment:
I don't think you are supposed to TryPeek and then assume when you call TryDequeue that you get the same item. In the intervening time another thread could have dequeued it.

Reply:
Some questions I have here are, is this really a queue model since we're not dealing with FIFO? Are the workid's guaranteed to be unique? If so couldn't we do this with a dictionary?

Reply:
Well, messages have to be processed in the order they came in, so it is in fact FIFO in this MQTT messaging model.

Reply:
I have removed the TryPeek approach and dequeuing it instead, if TryPeek returned the wrong workitem, we would have thrown a non-transient exception so it doesn't seem to matter whether or not we dequeue the message or not at that point, we can have the same functionality using dequeue and that should resolve the issue.